### DISCUSSION TITLE
Guacamole - unable to allocate a TLS value for currentThread

### DISCUSSION THREAD

**Discussion by jbufasttrack** [FreeRDP/FreeRDP#9717] [Q&A]
We're using Guacamole 1.5.4 which incorporates FreeRDP 2.11.4. When creating multiple connection attempts via RDP we can consistently get our setup into a state where FreeRDP throws the Major bug, unable to allocate a TLS value for currentThread error: https://github.com/FreeRDP/FreeRDP/blob/87557b1e4ef636e18a7dea58f0cae76fa5dff4fb/winpr/libwinpr/thread/thread.c#L439

The error occurs when we do exactly 59 connections. This results in a SSL_CTX_new failed error and after this every connection attempt throws the "unable to allocate a TLS value for currentThread" error. 

The only way to get out of this state is to restart the server / docker image. This gives you another 59 connections before throwing the same error again. 

We're currently looking at the code for both FreeRdp and Guacamole, but at the moment our prime suspect is that a change from Debian to Alpine for building dependencies in Guacamole might have changed some default limits / variables in FreeRDP. 

Have any others experienced the same behavior or have any explanations as to what might be causing and environment to throw the above errors? And also how to prevent this? 

We also have a thread open with the Guacamole team: https://www.mail-archive.com/user@guacamole.apache.org/msg12474.html as it's currently not entirely clear if this is a Guacamole or FreeRDP issue. 

Any assistance will be hugely appreciated.

**Comment by mfleisz** [comment]
@jbufasttrack could you do the following test with your setup:

- Could you add a log statement to the beginning of `initializeThreads` in `theads.c` and check if it is only called once during the apps whole lifetime.
- Could you log the result of `pthread_key_create` in `TlsAlloc` (in `tls.c`) so that we know why the call is actually failing

**Reply by jbufasttrack** [reply]
@mfleisz we actually just did a rollback to Guacamole 1.5.3 (we were running 1.5.4 before) and this has completely solved the issue. 
So, it would seem that this should be directed at the Guacamole team instead :)

**Reply by arulrajnet** [reply]
Does this issue fixed in Guacamole 1.5.5? Anyone knows?

**Reply by kxs-sindrakumar** [reply]
@arulrajnet did you ever find answer to this?

**Reply by xarem** [reply]
@arulrajnet @kxs-sindrakumar we can conform that it's **not** fixed in 1.5.5.

**Reply by Costya-Y** [reply]
We ended up rolling back to 1.5.3 as well.
